### PR TITLE
Implement Apple TV Platform

### PR DESCRIPTION
**Pull request by fuzzard**
## Description
Implement the AppleTV platform for ATV4/4k

## Motivation and Context
Well here we are, years in the making. May as well say thanks to everyone to get it this far, Memphiz, Davilla (Team Mrmc), Sylvain, Kambala, Phunkyfish, Pogar, and anyone else who has helped that ive missed.

Notes:
PR is not in buildable individual commits. We have broken it up like this for, hopefully, for easier review, particularly the non platform code that is touched. Before Merge, the intent is to squash most into a single commit.
Feel free to update Labels as you see relevant.

Todo:
Currently a lot of refactoring work is being done around darwin_embedded platform as a whole. This will obviously bleed into tvos, but will tackle it later. Feel free to comment on any architecture related issues, but keep in mind, that will most likely go on the todo list for now.

## How Has This Been Tested?
ATV dev's, and some external users who have helped with testing

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [X] **New feature** (non-breaking change which adds functionality)
- [X] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [X] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [X] All new and existing tests passed

### CODE REVIEW

**Review by Rechi — CHANGES_REQUESTED** [review]
You are not following the [code guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md).

**Comment by fuzzard** [comment]
Should be better now rechi. The 2 files that don't conform are 3rd party code we use for a gzip library

File: tools/depends/target/python3/darwin_embedded.patch [hunk 1]
```diff
@@ -197,6 +197,7 @@
      if (pid == 0) {
          /* Child process */
          /*
+
```
Comment (Rechi) [inline tools/depends/target/python3/darwin_embedded.patch]: Why is this line needed?

File: tools/depends/target/python3/darwin_embedded.patch [hunk 2]
```diff
@@ -214,7 +215,7 @@
          wchar_t *path = Py_DecodeLocale(execpath, &len);
 --- a/Modules/posixmodule.c
 +++ b/Modules/posixmodule.c
-@@ -197,6 +197,20 @@
+@@ -194,6 +194,20 @@
```
Comment (Rechi) [inline tools/depends/target/python3/darwin_embedded.patch]: Is the current range information incorrect?

File: tools/depends/target/python3/darwin_embedded.patch [hunk 3]
```diff
@@ -231,11 +232,11 @@
 +#  undef HAVE_WAIT
 +#  undef HAVE_WAIT3
 +#  undef HAVE_WAIT4
-+#  undef HAVE_WAITPID
++//#  undef HAVE_WAITPID
```
Comment (Rechi) [inline tools/depends/target/python3/darwin_embedded.patch]: Why patch commented `#  undef HAVE_WAITPID` in, instead of not patching this line in?

File: tools/depends/target/python3/darwin_embedded.patch [hunk 4]
```diff
@@ -231,11 +232,11 @@
...
  
  /*[clinic input]
  # one of the few times we lie about this name!
-@@ -475,6 +489,10 @@
+@@ -472,6 +486,10 @@
```
Comment (Rechi) [inline tools/depends/target/python3/darwin_embedded.patch]: Is the current range information incorrect?

File: tools/depends/target/python3/darwin_embedded.patch [hunk 5]
```diff
@@ -246,7 +247,7 @@
  /* Legacy wrapper */
  void
  PyOS_AfterFork(void)
-@@ -5040,6 +5058,8 @@
+@@ -5037,6 +5055,8 @@
```
Comment (Rechi) [inline tools/depends/target/python3/darwin_embedded.patch]: Is the current range information incorrect?

File: tools/depends/target/python3/darwin_embedded.patch [hunk 6]
```diff
@@ -255,7 +256,7 @@
      /* If we get here it's definitely an error */
  
      free_string_array(argvlist, argc);
-@@ -5116,6 +5136,12 @@
+@@ -5113,6 +5133,12 @@
```
Comment (Rechi) [inline tools/depends/target/python3/darwin_embedded.patch]: Is the current range information incorrect?

File: tools/depends/target/python3/darwin_embedded.patch [hunk 7]
```diff
@@ -268,7 +269,7 @@
      /* If we get here it's definitely an error */
  
      posix_path_error(path);
-@@ -5416,7 +5442,7 @@
+@@ -5413,7 +5439,7 @@
```
Comment (Rechi) [inline tools/depends/target/python3/darwin_embedded.patch]: Is the current range information incorrect?

File: tools/depends/target/python3/darwin_embedded.patch [hunk 8]
```diff
@@ -277,7 +278,7 @@
      if (pid == 0) {
          /* child: this clobbers and resets the import lock. */
          PyOS_AfterFork_Child();
-@@ -5447,7 +5473,7 @@
+@@ -5444,7 +5470,7 @@
```
Comment (Rechi) [inline tools/depends/target/python3/darwin_embedded.patch]: Is the current range information incorrect?

File: tools/depends/target/python3/darwin_embedded.patch (hunk 1 above)
Comment (fuzzard) [reply tools/depends/target/python3/darwin_embedded.patch]: was done by the diff command creating the patch

File: tools/depends/target/python3/darwin_embedded.patch (hunk 3 above)
Comment (fuzzard) [reply tools/depends/target/python3/darwin_embedded.patch]: at the time was just easier to update the patch by modifying the comment rather than rediff it all. I'll get it removed proper now

File: tools/depends/target/python3/darwin_embedded.patch (hunk 4 above)
Comment (fuzzard) [reply tools/depends/target/python3/darwin_embedded.patch]: Yes

File: tools/depends/target/python3/darwin_embedded.patch (hunk 5 above)
Comment (fuzzard) [reply tools/depends/target/python3/darwin_embedded.patch]: Yes

File: tools/depends/target/python3/darwin_embedded.patch (hunk 6 above)
Comment (fuzzard) [reply tools/depends/target/python3/darwin_embedded.patch]: Yes

File: tools/depends/target/python3/darwin_embedded.patch (hunk 7 above)
Comment (fuzzard) [reply tools/depends/target/python3/darwin_embedded.patch]: Yes

File: tools/depends/target/python3/darwin_embedded.patch (hunk 8 above)
Comment (fuzzard) [reply tools/depends/target/python3/darwin_embedded.patch]: Yes

File: tools/depends/target/python3/darwin_embedded.patch (hunk 2 above)
Comment (fuzzard) [reply tools/depends/target/python3/darwin_embedded.patch]: Yes, all ranges are correct. They are based off the preceding patches having been applied (5 patches) as laid out in the makefile.

File: tools/depends/target/python3/darwin_embedded.patch (hunk 1 above)
Comment (Rechi) [reply tools/depends/target/python3/darwin_embedded.patch]: It for sure wasn't, the above hunk applies cleanly without this additional line.

File: cmake/platform/darwin_embedded/tvos.cmake [hunk 9]
```diff
@@ -0,0 +1,3 @@
+set(PLATFORM_REQUIRED_DEPS OpenGLES)
+set(APP_RENDER_SYSTEM gles)
+list(APPEND PLATFORM_DEFINES -DTARGET_DARWIN_TVOS)
```
Comment (Rechi) [inline cmake/platform/darwin_embedded/tvos.cmake]: add an empty line at the end of file

File: tools/buildsteps/tvos/prepare-xbmc [hunk 10]
```diff
@@ -4,6 +4,6 @@ XBMC_PLATFORM_DIR=tvos
 
 #build binary addons before building xbmc...
 #make sure that binary_addons don't clean the native tools
-#here
+#here (e.x. on release builds where pathChanged always returns 1
```
Comment (Rechi) [inline tools/buildsteps/tvos/prepare-xbmc]: really?

File: xbmc/platform/darwin/tvos/Assets.xcassets/Assets.brandassets/Contents.json [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+      "idiom" : "tv",
+      "filename" : "topshelf.imageset",
+      "role" : "top-shelf-image"
+    }
+  ],
+  "info" : {
+    "version" : 1,
+    "author" : "xcode"
+  }
+}
```
Comment (Rechi) [inline xbmc/platform/darwin/tvos/Assets.xcassets/Assets.brandassets/Contents.json]: add an empty line at the end of file

File: xbmc/platform/darwin/tvos/Assets.xcassets/Assets.brandassets/topshelf.imageset/Contents.json [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+    {
+      "idiom" : "tv",
+      "scale" : "2x"
+    }
+  ],
+  "info" : {
+    "version" : 1,
+    "author" : "xcode"
+  }
+}
```
Comment (Rechi) [inline xbmc/platform/darwin/tvos/Assets.xcassets/Assets.brandassets/topshelf.imageset/Contents.json]: add an empty line at the end of file

File: xbmc/settings/DisplaySettings.cpp [hunk 13]
```diff
@@ -822,19 +831,14 @@ void CDisplaySettings::SettingOptionsPreferredStereoscopicViewModesFiller(Settin
...
+    defined(TARGET_WINDOWS_DESKTOP)
   std::vector<std::string> monitors;
-
-#if defined(HAVE_X11)
-  CWinSystemX11 *winSystem = dynamic_cast<CWinSystemX11*>(CServiceBroker::GetWinSystem());
-#elif defined(TARGET_DARWIN_OSX)
-  CWinSystemOSX *winSystem = dynamic_cast<CWinSystemOSX*>(CServiceBroker::GetWinSystem());
-#elif defined(TARGET_DARWIN_IOS)
-  CWinSystemIOS *winSystem = dynamic_cast<CWinSystemIOS*>(CServiceBroker::GetWinSystem());
-#endif
```
Comment (Rechi) [inline xbmc/settings/DisplaySettings.cpp]: The refactoring should be moved into a separate commit.

File: xbmc/platform/darwin/ios-common/DarwinNSUserDefaults.h [hunk 14]
```diff
@@ -1,28 +1,50 @@
...
+ *
+ *  This Program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with MrMC; see the file COPYING.  If not, see
+ *  <http://www.gnu.org/licenses/>.
+ *
```
Comment (Rechi) [inline xbmc/platform/darwin/ios-common/DarwinNSUserDefaults.h]: Why?

File: xbmc/platform/darwin/ios-common/DarwinNSUserDefaults.mm [hunk 15]
```diff
@@ -2,27 +2,70 @@
...
+ *
+ *  This Program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with MrMC; see the file COPYING.  If not, see
+ *  <http://www.gnu.org/licenses/>.
+ *
```
Comment (Rechi) [inline xbmc/platform/darwin/ios-common/DarwinNSUserDefaults.mm]: Why?

File: cmake/scripts/darwin_embedded/Install.cmake [hunk 16]
```diff
@@ -80,6 +80,21 @@ set_target_properties(${APP_NAME_LC} PROPERTIES XCODE_ATTRIBUTE_CODE_SIGN_IDENTI
...
+  endif()
+  add_custom_target(binary-addons
+    COMMAND $(MAKE) -C ${CMAKE_SOURCE_DIR}/tools/depends/target/binary-addons clean
+    COMMAND $(MAKE) -C ${CMAKE_SOURCE_DIR}/tools/depends/target/binary-addons VERBOSE=1 V=99
+          INSTALL_PREFIX="${CMAKE_BINARY_DIR}/addons" CROSS_COMPILING=yes ${_addons})
+  if(ENABLE_XCODE_ADDONBUILD)
+    add_dependencies(${APP_NAME_LC} binary-addons)
+  endif()
+  unset(_addons)
+endif()
```
Comment (Rechi) [inline cmake/scripts/darwin_embedded/Install.cmake]: Why only add this for tvOS?

**Review by Rechi — COMMENTED** [review]
For the most commits in this PR I don't know what changes to expect as it is the directory path of the changed files.

File: xbmc/platform/darwin/ios-common/DarwinNSUserDefaults.h (hunk 14 above)
Comment (fuzzard) [reply xbmc/platform/darwin/ios-common/DarwinNSUserDefaults.h]: Was a backport from mrmc codebase, copied the licensing as is from that

Comment (Rechi) [reply xbmc/platform/darwin/ios-common/DarwinNSUserDefaults.h]: Please use SPDX licensing header.

File: cmake/scripts/darwin_embedded/Install.cmake (hunk 16 above)
Comment (fuzzard) [reply cmake/scripts/darwin_embedded/Install.cmake]: Was put in the cmakebuildsys PR earlier, but was asked to be pulled out for "discussion".
Because of timelines, i felt it easier to attribute to tvos only, and if its desired for ios, i can remove the platform check at a later date.

File: xbmc/settings/DisplaySettings.cpp (hunk 13 above)
Comment (fuzzard) [reply xbmc/settings/DisplaySettings.cpp]: Done.

**Comment by fuzzard** [comment]
> For the most commits in this PR I don't know what changes to expect as it is the directory path of the changed files.

Not sure what you mean by this? is this for the cmake treedata split for tvos/ios

File: cmake/scripts/darwin_embedded/Install.cmake (hunk 16 above)
Comment (Rechi) [reply cmake/scripts/darwin_embedded/Install.cmake]: It wasn't related to the previous PR and still isn't to this one. If you want this feature, open a separate PR, which implements it for all platforms.

Comment (fuzzard) [reply cmake/scripts/darwin_embedded/Install.cmake]: Its a feature to ease building for the apple tv platform, how is that not relevant here? Just because it could be used for ios is not the point. We are working towards updating ios stuff as well, and thats all still to come, why not add a feature for THIS platform now?

**Comment by Rechi** [comment]
E.g. the commit message `[tvOS] system`, `[tvOS] xbmc` or `[tvOS] xbmc/addons`. I have no idea what it changes, without looking at it.
Actually I see no reason to split the changes for tvOS per xbmc subdirectory, because they do not work / compile stand-alone.

**Comment by fuzzard** [comment]
```Notes:
PR is not in buildable individual commits. We have broken it up like this for, hopefully, for easier review, particularly the non platform code that is touched. Before Merge, the intent is to squash most into a single commit.```

Thats a quote from the PR message. The intent is for reviewers to see non platform code at a glance for reviewers relevant areas of expertise. It will not merge with this many commits, it WILL be squashed. If you want it squashed now, thats fine by me, it will be a 3 commits.

File: cmake/scripts/darwin_embedded/Install.cmake (hunk 16 above)
Comment (Rechi) [reply cmake/scripts/darwin_embedded/Install.cmake]: It's not a feature for tvOS only and also not for tvOS and iOS only, it is a general feature.
The reason I ask for doing it right from the beginning is, that otherwise it might get left behind.

Comment (fuzzard) [reply cmake/scripts/darwin_embedded/Install.cmake]: Honestly i dont have time to vet it for macos right now. I know it works for ios, but again, i PR'd this early knowing there would be feedback, and i understand where you are coming from regarding "forgotten" catch up features for other platforms, but the point is to get this in for Alpha no? Not everything will be perfect, and theres plenty of todo's. 

I would rather put it through in its known tested state, and when time permits and proper testing is done, open up the platform check to others.

**Comment by fuzzard** [comment]
Jenkins build this please

**Comment by fuzzard** [comment]
jenkins build this please

**Comment by fuzzard** [comment]
Jenkins build this please

**Comment by fuzzard** [comment]
Jenkins build this please

File: xbmc/platform/darwin/tvos/filesystem/TVOSDirectory.cpp [hunk 17]
```diff
@@ -0,0 +1,130 @@
...
+{
+  std::string rootpath = CSpecialProtocol::TranslatePath(url);
+  size_t found = rootpath.find(CTVOSFileUtils::GetUserHomeDirectory());
+  if (found == std::string::npos)
+    return false;
+
+  return true;
+}
+
+bool CTVOSDirectory::GetDirectory(const CURL& url, CFileItemList& items)
```
Comment (kambala-decapitator) [inline xbmc/platform/darwin/tvos/filesystem/TVOSDirectory.cpp]: auto where possible

File: xbmc/platform/darwin/tvos/filesystem/TVOSFile.cpp [hunk 18]
```diff
@@ -0,0 +1,274 @@
...
+
+#include "platform/darwin/ios-common/DarwinNSUserDefaults.h"
+#include "platform/posix/filesystem/PosixFile.h"
+
+#include <sys/stat.h>
+
+using namespace XFILE;
+
+
+CTVOSFile::CTVOSFile() : m_position(-1), m_pFallbackFile(nullptr)
```
Comment (kambala-decapitator) [inline xbmc/platform/darwin/tvos/filesystem/TVOSFile.cpp]: set default values in class declaration, = default instead of empty body

File: xbmc/platform/darwin/tvos/filesystem/TVOSFile.cpp [hunk 19]
```diff
@@ -0,0 +1,274 @@
...
+
+  void* lpBufInternal = nullptr;
+
+  if (m_position > 0 && m_position == GetLength())
+    return 0; // simulate read 0 bytes on EOF
+
+  if (CDarwinNSUserDefaults::GetKeyDataFromPath(m_url.Get(), lpBufInternal,
+                                                uiBufSize)) // read size of file
+  {
+    lpBufInternal = static_cast<void*>(new char[uiBufSize]);
```
Comment (kambala-decapitator) [inline xbmc/platform/darwin/tvos/filesystem/TVOSFile.cpp]: use smart pointer

File: xbmc/platform/darwin/tvos/filesystem/TVOSFileUtils.mm [hunk 20]
```diff
@@ -0,0 +1,41 @@
...
+#import "utils/URIUtils.h"
+
+#import <mutex>
+
+#import <Foundation/Foundation.h>
+
+const char* CTVOSFileUtils::GetUserHomeDirectory()
+{
+  static std::string appHomeFolder;
+  if (appHomeFolder.empty())
```
Comment (kambala-decapitator) [inline xbmc/platform/darwin/tvos/filesystem/TVOSFileUtils.mm]: use std::call_once instead

File: xbmc/platform/darwin/tvos/input/LibInputRemote.mm [hunk 21]
```diff
@@ -0,0 +1,156 @@
...
+#import "windowing/tvos/WinSystemTVOS.h"
+
+#import "platform/darwin/tvos/XBMCController.h"
+#import "platform/darwin/tvos/input/LibInputHandler.h"
+#import "platform/darwin/tvos/input/LibInputSettings.h"
+
+#import <Foundation/Foundation.h>
+#import <UIKit/UIEvent.h>
+
+@implementation TVOSLibInputRemote
```
Comment (kambala-decapitator) [inline xbmc/platform/darwin/tvos/input/LibInputRemote.mm]: auto where possible

File: xbmc/platform/darwin/tvos/input/LibInputRemote.mm [hunk 22]
```diff
@@ -0,0 +1,156 @@
...
+{
+  switch (receivedEvent.subtype)
+  {
+    case UIEventSubtypeRemoteControlTogglePlayPause:
+      KODI::MESSAGING::CApplicationMessenger::GetInstance().PostMsg(
+          TMSG_GUI_ACTION, WINDOW_INVALID, -1,
+          static_cast<void*>(new CAction(ACTION_PLAYER_PLAYPAUSE)));
+      break;
+    case UIEventSubtypeRemoteControlPlay:
+      [g_xbmcController.inputHandler sendButtonPressed:13];
```
Comment (kambala-decapitator) [inline xbmc/platform/darwin/tvos/input/LibInputRemote.mm]: isn't there some enum instead of these magic numbers?

File: xbmc/platform/darwin/tvos/input/LibInputSettings.h [hunk 23]
```diff
@@ -0,0 +1,20 @@
...
+ *  Copyright (C) 2019- Team Kodi
+ *  This file is part of Kodi - https://kodi.tv
+ *
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *  See LICENSES/README.md for more information.
+ */
+
+#import <Foundation/Foundation.h>
+
+extern float const REPEATED_KEYPRESS_DELAY_S;
```
Comment (kambala-decapitator) [inline xbmc/platform/darwin/tvos/input/LibInputSettings.h]: 1. both should have type NSTimeInterval
2. why are they defined here, but actually used in another file?

File: xbmc/platform/darwin/tvos/input/LibInputTouch.mm [hunk 24]
```diff
@@ -0,0 +1,647 @@
...
+      break;
+
+    // auto-repeat keys
+    case UIPressTypeUpArrow:
+    case UIPressTypeDownArrow:
+    case UIPressTypeLeftArrow:
+    case UIPressTypeRightArrow:
+      break;
+
+    default:
```
Comment (kambala-decapitator) [inline xbmc/platform/darwin/tvos/input/LibInputTouch.mm]: if all possible cases are covered, I suggest to remove `default`

File: xbmc/platform/darwin/tvos/input/LibInputTouch.mm [hunk 25]
```diff
@@ -0,0 +1,647 @@
...
+  auto pan = [[UIPanGestureRecognizer alloc] initWithTarget:self action:@selector(handlePan:)];
+  pan.delegate = self;
+  [g_xbmcController.glView addGestureRecognizer:pan];
+}
+
+- (void)createTapGesturecognizers
+{
+  // tap side of siri remote pad
+  for (auto t : {
+         std::make_tuple(UIPressTypeUpArrow, @selector(tapUpArrowPressed:),
```
Comment (kambala-decapitator) [inline xbmc/platform/darwin/tvos/input/LibInputTouch.mm]: should be possible to use `{}` instead of std::make_tuple(), at least for all params after the first

File: xbmc/platform/darwin/tvos/input/LibInputTouch.mm [hunk 26]
```diff
@@ -0,0 +1,647 @@
...
+    [g_xbmcController.inputHandler sendButtonPressed:4];
+
+  [g_xbmcController.inputHandler.inputRemote startRemoteTimer];
+}
+
+#pragma mark - Pan
+
+- (IBAction)handlePan:(UIPanGestureRecognizer*)sender
+{
+  if (!g_xbmcController.inputHandler.inputRemote.remoteIdleState)
```
Comment (kambala-decapitator) [inline xbmc/platform/darwin/tvos/input/LibInputTouch.mm]: early return for this and next if

File: xbmc/platform/darwin/tvos/input/LibInputTouch.mm [hunk 27]
```diff
@@ -0,0 +1,647 @@
...
+                  if (ABS(velocityX) > minVelocity && [self shouldFastScroll])
+                  {
+                    [g_xbmcController.inputHandler sendButtonPressed:11];
+                  }
+                  m_lastGesturePoint = gesturePoint;
+                }
+                break;
+              }
+              default:
+              {
```
Comment (kambala-decapitator) [inline xbmc/platform/darwin/tvos/input/LibInputTouch.mm]: omit braces

File: xbmc/utils/SystemInfo.cpp [hunk 28]
```diff
@@ -1258,6 +1274,10 @@ std::string CSysInfo::GetBuildTargetPlatformVersionDecoded(void)
 #elif defined(TARGET_DARWIN_IOS)
   return StringUtils::Format("version %d.%d.%d", (__IPHONE_OS_VERSION_MIN_REQUIRED / 10000) % 100,
                              (__IPHONE_OS_VERSION_MIN_REQUIRED / 100) % 100, __IPHONE_OS_VERSION_MIN_REQUIRED % 100);
+#elif defined(TARGET_DARWIN_TVOS)
```
Comment (kambala-decapitator) [inline xbmc/utils/SystemInfo.cpp]: iOS and tvOS can be merged into one using `XSTR_MACRO`

File: xbmc/windowing/tvos/OSScreenSaverTVOS.mm [hunk 29]
```diff
@@ -0,0 +1,25 @@
...
+ *
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *  See LICENSES/README.md for more information.
+ */
+
+#import "OSScreenSaverTVOS.h"
+
+#import "platform/darwin/tvos/XBMCController.h"
+
+COSScreenSaverTVOS::COSScreenSaverTVOS()
```
Comment (kambala-decapitator) [inline xbmc/windowing/tvos/OSScreenSaverTVOS.mm]: = default

File: xbmc/windowing/tvos/VideoSyncTVos.h [hunk 30]
```diff
@@ -0,0 +1,44 @@
...
+
+#include "guilib/DispResource.h"
+#include "windowing/VideoSync.h"
+
+class CWinSystemTVOS;
+
+class CVideoSyncTVos : public CVideoSync, IDispResource
+{
+public:
+  CVideoSyncTVos(void* clock, CWinSystemTVOS& winSystem) : CVideoSync(clock), m_winSystem(winSystem)
```
Comment (kambala-decapitator) [inline xbmc/windowing/tvos/VideoSyncTVos.h]: method definition should not be in class declaration

File: xbmc/windowing/tvos/WinSystemTVOS.mm [hunk 31]
```diff
@@ -0,0 +1,438 @@
...
+
+#define CONST_HDMI "HDMI"
+
+// if there was a devicelost callback
+// but no device reset for 3 secs
+// a timeout fires the reset callback
+// (for ensuring that e.x. AE isn't stuck)
+constexpr uint32_t LOST_DEVICE_TIMEOUT_MS{3000};
+
+// TVOSDisplayLinkCallback is declared in the lower part of the file
```
Comment (kambala-decapitator) [inline xbmc/windowing/tvos/WinSystemTVOS.mm]: probably defined, not declared? :)

File: xbmc/windowing/tvos/WinSystemTVOS.mm [hunk 32]
```diff
@@ -0,0 +1,438 @@
...
+
+struct CADisplayLinkWrapper
+{
+  CADisplayLink* impl;
+  TVOSDisplayLinkCallback* callbackClass;
+};
+
+std::unique_ptr<CWinSystemBase> CWinSystemBase::CreateWinSystem()
+{
+  std::unique_ptr<CWinSystemBase> winSystem(new CWinSystemTVOS());
```
Comment (kambala-decapitator) [inline xbmc/windowing/tvos/WinSystemTVOS.mm]: use `std::make_unique` instead

File: xbmc/windowing/tvos/WinSystemTVOS.mm [hunk 33]
```diff
@@ -0,0 +1,438 @@
...
+  return true;
+}
+
+bool CWinSystemTVOS::GetScreenResolution(int* w, int* h, double* fps, int screenIdx)
+{
+  *w = [g_xbmcController.displayManager getScreenSize].width;
+  *h = [g_xbmcController.displayManager getScreenSize].height;
+  *fps = [g_xbmcController.displayManager getDisplayRate];
+
+  CLog::Log(LOGDEBUG, "Current resolution Screen: {} with {} x {} @  {}", screenIdx, *w, *h, *fps);
```
Comment (kambala-decapitator) [inline xbmc/windowing/tvos/WinSystemTVOS.mm]: extra space after @

File: xbmc/windowing/tvos/WinSystemTVOS.mm [hunk 34]
```diff
@@ -0,0 +1,438 @@
...
+
+  //now just fill in the possible resolutions for the attached screens
+  //and push to the resolution info vector
+  FillInVideoModes(screenIdx);
+}
+
+void CWinSystemTVOS::FillInVideoModes(int screenIdx)
+{
+  // Potential refresh rates
+  std::vector<float> supportedDispRefreshRates = {23.976, 24.000, 25.000, 29.970,
```
Comment (kambala-decapitator) [inline xbmc/windowing/tvos/WinSystemTVOS.mm]: add `f` suffix to all values

**Review by kambala-decapitator — CHANGES_REQUESTED** [review]

File: xbmc/platform/darwin/ios-common/IOSKeyboardView.mm [hunk 35]
```diff
@@ -13,100 +13,111 @@
...
+#pragma clang diagnostic ignored "-Wobjc-designated-initializers"
+- (instancetype)initWithFrame:(CGRect)frame
+#pragma clang diagnostic pop
 {
-  _frame = frame;
-  if([NSThread currentThread] != [NSThread mainThread])
-  {
-    [self performSelectorOnMainThread:@selector(initWithFrameInternal) withObject:nil  waitUntilDone:YES];
-  }
+  // @todo in fact, it's client's responsibility to init on main thread
```
[the remark this answers is not in the document]
Comment (fuzzard) [reply xbmc/platform/darwin/ios-common/IOSKeyboardView.mm]: All the keyboard stuff we were reworking. In that branch i gave you a few weeks ago regarding the ios keyboard rework. From memory, it fixed this, and removed that todo.

Was waiting on this going into master before completing the tvos part of that however

File: xbmc/platform/darwin/tvos/input/LibInputSettings.h (hunk 23 above)
Comment (fuzzard) [reply xbmc/platform/darwin/tvos/input/LibInputSettings.h]: Only reason they were thrown in here was they were some magic number hidden away in ther original location. I started to strip anything that may need to be user changable into Settings.

Happy to move out from here, as after the refactor, i dont think it ended up making sense to make it an end user setting.

File: xbmc/platform/darwin/tvos/filesystem/TVOSFile.cpp (hunk 19 above)
Comment (fuzzard) [reply xbmc/platform/darwin/tvos/filesystem/TVOSFile.cpp]: discussed this one with you. At this stage i'll leave it as is. May be able to change it with some refactoring. Will look into it at a later date.

File: xbmc/platform/darwin/tvos/TopShelf/ServiceProvider.m [hunk 36]
```diff
@@ -0,0 +1,121 @@
...
+@implementation ServiceProvider
+
+#pragma mark - TVTopShelfProvider protocol
+
+- (TVTopShelfContentStyle)topShelfStyle
+{
+  return TVTopShelfContentStyleSectioned;
+}
+
+- (NSArray<TVContentItem*>*)topShelfItems
```
[the remark this answers is not in the document]
Comment (fuzzard) [reply xbmc/platform/darwin/tvos/TopShelf/ServiceProvider.m]: went the __auto_type route. Was just the easier option than converting to mm

**Comment by fuzzard** [comment]
Jenkins build this please

**Comment by fuzzard** [comment]
Jenkins build this please

**Comment by fuzzard** [comment]
jenkins build this please

**Comment by fuzzard** [comment]
Jenkins build this please

**Comment by fuzzard** [comment]
Jenkins build this please

**Comment by fuzzard** [comment]
Jenkins build this please

**Comment by fuzzard** [comment]
Jenkins build this please

**Comment by fuzzard** [comment]
Jenkins build this please

File: docs/README.tvOS.md [hunk 37]
```diff
@@ -103,140 +109,141 @@ make -j$(getconf _NPROCESSORS_ONLN)
...
 ```
-make -C tools/depends/target/binary-addons
+make -C tools/depends/target/cmakebuildsys BUILD_DIR=$HOME/kodi-build
 ```
 
-Build specific add-ons:
-```
-make -C tools/depends/target/binary-addons ADDONS="audioencoder.flac pvr.vdr.vnsi audiodecoder.snesapu"
-```
+**TIP:** BUILD_DIR can be omitted, and project will be created in $HOME/kodi/build
```
Comment (chkpnt) [inline docs/README.tvOS.md]: I just tried it out: The project has been created in `./build`.

Comment (fuzzard) [reply docs/README.tvOS.md]: Not sure I see the problem?

All conventions in the readme follow on from prior steps, so if you set it up as described, Kodi source gets cloned into $HOME/Kodi, which is where you execute the cmakebuildsys command. This is then going to be ./build, as you are in the directory, or long form $HOME/Kodi/build

**Comment by fuzzard** [comment]
Jenkins build this please

File: docs/README.tvOS.md (hunk 37 above)
Comment (chkpnt) [reply docs/README.tvOS.md]: You are right, I'm sorry. I've missunderstood the sentence, thought `$HOME/kodi/build` was hard coded somewere.

**Comment by fuzzard** [comment]
Jenkins build this please

**Comment by fuzzard** [comment]
Jenkins build this please

**Comment by Memphiz** [comment]
Yay - congrats - thx and well done to all involved!

**Comment by pogarek** [comment]
Indeed. Well done and congrats. Finally it happened !
